### PR TITLE
feat(gateway): separate logs, metrics and traces settings, start logging

### PR DESCRIPTION
**Pull request by pimeys**
This adds logs into strategic places: subgraph response, federated graph response and gateway response. Normal stuff in debug path, errors in errors. Will not work with Grafbase analytics yet.

Logs, traces and metrics can be configured separately, or in one block in the global settings. Logs will be sent into opentelemetry.

### CODE REVIEW

File: engine/crates/telemetry/src/TelemetryConfig [hunk 1]
```diff
@@ -0,0 +1 @@
+
```
Comment (tomhoule) [inline engine/crates/telemetry/src/TelemetryConfig:1]: Empty file expected?

File: gateway/crates/gateway-binary/tests/telemetry/metrics/.request.rs.pending-snap
Comment (tomhoule) [inline gateway/crates/gateway-binary/tests/telemetry/metrics/.request.rs.pending-snap:1]: pending snapshot

File: gateway/crates/gateway-binary/tests/telemetry/metrics/.operation.rs.pending-snap
Comment (tomhoule) [inline gateway/crates/gateway-binary/tests/telemetry/metrics/.operation.rs.pending-snap:1]: pending snapshot

File: engine/crates/engine-v2/src/engine.rs [hunk 2]
```diff
@@ -192,25 +192,31 @@ impl<R: Runtime> Engine<R> {
...
+
             if let Some(OperationMetadata {
                 ty,
                 name,
                 normalized_query,
                 normalized_query_hash,
             }) = operation_metadata
             {
-                span.record_gql_request(GqlRequestAttributes {
+                tracing::Span::current().record_gql_request(GqlRequestAttributes {
```
Comment (Finistere) [inline engine/crates/engine-v2/src/engine.rs:211]: I preferred specifying explicitly the `span` IMO

File: engine/crates/engine-v2/src/engine.rs [hunk 3]
```diff
@@ -221,13 +227,27 @@ impl<R: Runtime> Engine<R> {
...
+            if status.is_success() {
+                tracing::Span::current().record_gql_status(status, elapsed, None);
+                tracing::debug!(target: GRAFBASE_TARGET, "router response")
+            } else {
+                let message = response
+                    .first_error_message()
+                    .map(|s| s.into_owned())
+                    .unwrap_or_else(|| String::from("router error"));
+
+                tracing::Span::current().record_gql_status(status, elapsed, Some(message.clone()));
```
Comment (Finistere) [inline engine/crates/engine-v2/src/engine.rs:243]: For tracing, I don't find it obvious that we want the first error message and only that right now. I would not touch it as of now. So `message` should only be used for logs

File: engine/crates/engine-v2/src/engine.rs [hunk 4]
```diff
@@ -221,13 +227,27 @@ impl<R: Runtime> Engine<R> {
...
                     },
-                    start.elapsed(),
+                    elapsed,
                 );
             }
-            span.record_gql_status(status);
+
+            if status.is_success() {
+                tracing::Span::current().record_gql_status(status, elapsed, None);
+                tracing::debug!(target: GRAFBASE_TARGET, "router response")
```
Comment (Finistere) [inline engine/crates/engine-v2/src/engine.rs:236]:
```suggestion
                tracing::debug!(target: GRAFBASE_TARGET, "gateway request")
```

File: engine/crates/engine-v2/src/plan/mod.rs [hunk 5]
```diff
@@ -0,0 +1,137 @@
+use std::sync::Arc;
```
Comment (Finistere) [inline engine/crates/engine-v2/src/plan/mod.rs:1]: incorrect rebase?

File: engine/crates/engine-v2/src/plan/planning/collect.rs [hunk 6]
```diff
@@ -0,0 +1,680 @@
+use id_newtypes::IdRange;
```
Comment (Finistere) [inline engine/crates/engine-v2/src/plan/planning/collect.rs:1]: rebase error?

File: engine/crates/engine-v2/src/sources/graphql/Field [hunk 7]
```diff
@@ -0,0 +1 @@
+
```
Comment (Finistere) [inline engine/crates/engine-v2/src/sources/graphql/Field:1]: ?

File: engine/crates/engine-v2/src/response/write/mod.rs [hunk 8]
```diff
@@ -368,6 +368,20 @@ impl ResponsePart {
             inner: Rc::new(RefCell::new(self)),
         }
     }
+
+    pub fn first_error_message(&self) -> Option<String> {
+        self.errors.first().map(|s| s.message.to_string())
+    }
+
+    /// True if this request was blocked by the query plan. Avoid logging any request details if this is true.
+    pub fn blocked_in_planning(&self) -> bool {
```
Comment (Finistere) [inline engine/crates/engine-v2/src/response/write/mod.rs:377]: What you want is this:

```
pub fn subgraph_errors(&self) -> impl Iterator<Item = &GraphqlError> + '_ {
  self.errors.iter().filter(|e| matches!(e.code ErrorCode::SubgraphError| ErrorCode::SubgraphInvalidResponseError |ErrorCode::SubgraphRequestError))
}
```

And then use the first message of this to decide what to do from it.

File: engine/crates/engine-v2/src/sources/graphql/deserialize/errors.rs [hunk 9]
```diff
@@ -2,7 +2,7 @@ use serde::{de::DeserializeSeed, Deserializer};
 
 use crate::response::{ErrorCode, GraphqlError, ResponseKeys, ResponsePartMut, ResponsePath, UnpackedResponseEdge};
 
-pub(super) trait GraphqlErrorsSeed<'a> {
+pub trait GraphqlErrorsSeed<'a> {
```
Comment (Finistere) [inline engine/crates/engine-v2/src/sources/graphql/deserialize/errors.rs:5]: ?

File: engine/crates/integration-tests/src/federation/builder/mock.rs [hunk 10]
```diff
@@ -0,0 +1,102 @@
+use std::{
```
Comment (Finistere) [inline engine/crates/integration-tests/src/federation/builder/mock.rs:1]: rebase?

File: engine/crates/telemetry/src/otel/logs.rs [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+                .build()
+        };
+
+        builder = builder.with_log_processor(processor);
+    }
+
+    Ok(Some(builder.build()))
+}
+
+#[cfg(not(feature = "otlp"))]
```
Comment (Finistere) [inline engine/crates/telemetry/src/otel/logs.rs:50]: Using `#[cfg(feature = "otlp")]` on the `if` is better IMO, `not` condition are a pain to deal with rust-analyzer. they're always hidden. Please avoid them

File: engine/crates/telemetry/src/span.rs [hunk 12]
```diff
@@ -35,9 +35,28 @@ pub trait GqlRecorderSpanExt {
     fn record_gql_request(&self, attributes: GqlRequestAttributes<'_>);
     /// Record GraphQL response attributes in the span
     fn record_gql_response(&self, attributes: GqlResponseAttributes);
+    /// Record subgraph response attributes in the span
+    fn record_subgraph_response(&self, attributes: SubgraphResponseAttributes);
+
+    fn record_gql_error(&self, error: String);
+    fn record_gql_duration(&self, duration: Duration);
+
+    fn record_gql_status(&self, status: GraphqlResponseStatus, duration: Duration, error: Option<String>) {
```
Comment (Finistere) [inline engine/crates/telemetry/src/span.rs:44]: You shouldn't add any errors in the traces yet. It's unclear how we'll process them and we'll certainly want more than just the first error message.

File: engine/crates/telemetry/src/span/gql.rs [hunk 13]
```diff
@@ -30,6 +35,8 @@ impl GqlRequestSpan {
             "gql.response.field_errors_count"  = Empty,
             "gql.response.data_is_null"  = Empty,
             "gql.response.request_errors_count"  = Empty,
+            "gql.response.latency_ms" = Empty,
```
Comment (Finistere) [inline engine/crates/telemetry/src/span/gql.rs:38]: There is no point in adding `latency_ms` in a span, the span is _already_ an interval

Comment (Finistere) [reply engine/crates/telemetry/src/span/gql.rs:38]: If you want it in logs, add it to the message OR the log message attributes when creating the `tracing` event. not the span.

File: engine/crates/telemetry/src/tower.rs [hunk 14]
```diff
@@ -144,31 +144,57 @@ where
...
+                        gql_status,
                         client,
                     },
                     latency,
                 );
-                this.span.record_response(response);
+
+                match gql_status {
+                    Some(status) if status.is_success() => {
+                        Span::current().record_gql_status(status, latency, None);
```
Comment (Finistere) [inline engine/crates/telemetry/src/tower.rs:178]: why `Span::current()` when you have `this.span`?...

File: engine/crates/telemetry/src/gql_response_status.rs [hunk 15]
```diff
@@ -90,3 +88,28 @@ impl headers::Header for GraphqlResponseStatus {
         values.extend(Some(GraphqlResponseStatus::encode(self).try_into().unwrap()))
     }
 }
+
+#[derive(Clone, Copy, Debug)]
+pub enum SubgraphResponseStatus {
```
Comment (Finistere) [inline engine/crates/telemetry/src/gql_response_status.rs:93]: would rename `GraphqlRequestStatus` and re-use it for the top level gateway request span

**Review by Finistere — APPROVED** [review]

**Pull request merged**